### PR TITLE
Added EventHandler & GwtEvent class

### PR DESCRIPTION
**Pull request by FrankHossfeld**
Hi,
I have created the repo 'gwt-event-dom' which contains the dom event classes and interfaces. (everything from c.g.g.event.dom) and added gwt-events as dependency. I missed the GwtEvent and the EventHandler class in this repo. So I created a fork and added both. 

If you think, that's ok, it would be nice if you can merge the PR into the master branch.

Frank

### CODE REVIEW

File: src/main/java/org/gwtproject/event/shared/GwtEvent.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    assertLive();
+    return super.getSource();
+  }
+
+  /**
+   * Asserts that the event still should be accessed. All events are considered
+   * to be "dead" after their original handler manager finishes firing them. An
+   * event can be revived by calling {@link GwtEvent#revive()}.
+   */
+  protected void assertLive() {
```
Comment (tbroyer) [inline src/main/java/org/gwtproject/event/shared/GwtEvent.java]: If this, `isLive`, `kill`, and `revive` are **really** needed for DOM events, then I'd just add them to `DomEvent`. But I actually think this is just legacy from the time where the event was `window.event` in IE and there would only be one _live_ event at a given time.

File: src/main/java/org/gwtproject/event/shared/GwtEvent.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+  /**
+   * Revives the event. Used when recycling event instances.
+   */
+  protected void revive() {
+    dead = false;
+    setSource(null);
+  }
+
+  void overrideSource(Object source) {
```
Comment (tbroyer) [inline src/main/java/org/gwtproject/event/shared/GwtEvent.java]: This is never called, and inaccessible to other modules.

File: src/main/java/org/gwtproject/event/shared/GwtEvent.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  /**
+   * Type class used to register events with the {@link HandlerManager}.
+   * <p>
+   * Type is parameterized by the handler type in order to make the addHandler
+   * method type safe.
+   * </p>
+   * 
+   * @param <H> handler type
+   */
+  public static class Type<H> extends Event.Type<H> {
```
Comment (tbroyer) [inline src/main/java/org/gwtproject/event/shared/GwtEvent.java]: That type was only kept for backwards compatibility in GWT2. Given that we're killing backwards compat' by changing the package name, better not bring it here.

File: src/main/java/org/gwtproject/event/shared/GwtEvent.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2009 Google Inc.
```
Comment (tbroyer) [inline src/main/java/org/gwtproject/event/shared/GwtEvent.java]: Needs to be updated to “The GWT Project Authors”

**Review by tbroyer — COMMENTED** [review]
Those are part of the legacy that I had eliminated (see how the logical events directly extend `Event`); I'm 99% sure we don't need them.

File: src/main/java/org/gwtproject/event/shared/GwtEvent.java (hunk 4 above)
Comment (FrankHossfeld) [reply src/main/java/org/gwtproject/event/shared/GwtEvent.java]: Done

File: src/main/java/org/gwtproject/event/shared/GwtEvent.java (hunk 2 above)
Comment (FrankHossfeld) [reply src/main/java/org/gwtproject/event/shared/GwtEvent.java]: I will remove it.

File: src/main/java/org/gwtproject/event/shared/GwtEvent.java (hunk 1 above)
Comment (FrankHossfeld) [reply src/main/java/org/gwtproject/event/shared/GwtEvent.java]: Will move the method assertLive from GwtEvent to DomEvent and remove the method in the class GwtEvent

Comment (FrankHossfeld) [reply src/main/java/org/gwtproject/event/shared/GwtEvent.java]: Removing assertLive() in GwtEvent will cause an error in the method getSource

```

  @Override
  public Object getSource() {
    assertLive();
    return super.getSource();
  }
``` 
I will remove the method call. Do you think, that's ok?

**Comment by FrankHossfeld** [comment]
You are right, DomEvent can extend Event. So, there is no need to use GwtEvent.

And, @you are right. ```isLive```, ```kiil``` and ```revive``` are never called in dom-event. Removing these methods and deleting the ```assertLive```-method does not have an impact.

I remove the GwtEvent class.

**Comment by jnehlmeier** [comment]
> Those are part of the legacy that I had eliminated (see how the logical events directly extend Event); I'm 99% sure we don't need them.

@tbroyer I don't think `GwtEvent` / `EventHandler` can go away in a first release, given these new libraries should only require you to update imports if possible. Instead keep it and mark it deprecated if removal is desired in the future.

There is plenty of code extending `GwtEvent` directly: https://github.com/search?p=1&q=%22extends+GwtEvent%22&type=Code&utf8=✓

**Comment by tbroyer** [comment]
> […] given these new libraries should only require you to update imports if possible.

**Or** you could first migrate from `com.google.gwt.event` to `com.google.web.bindery.event` and then to `org.gwtproject.event`.

`com.google.gwt.event.` use has been discouraged for **7 years now,** it's time to move on!
https://github.com/gwtproject/gwt/commit/7669bfd3c7aaac0da0b74a00f1bd53008eca702b
> The GWT EventBus et al cannot be deprecated unless / until
> ActivityManager and friends move to com.google.web.bindery.

* [`EventBus`](http://www.gwtproject.org/javadoc/latest/com/google/gwt/event/shared/EventBus.html): “extends `com.google.web.bindery.event.shared.EventBus` for legacy compatibility.”
* [`SimpleEventBus`](http://www.gwtproject.org/javadoc/latest/com/google/gwt/event/shared/SimpleEventBus.html): “wraps `com.google.web.bindery.event.shared.SimpleEventBus` for legacy compatibility.”
* [`HandlerRegistration`](http://www.gwtproject.org/javadoc/latest/com/google/gwt/event/shared/HandlerRegistration.html): “extends `com.google.bindery.event.shared.HandlerRegistration` for legacy compatibility.”
* [`GwtEvent`](http://www.gwtproject.org/javadoc/latest/com/google/gwt/event/shared/GwtEvent.html): “There is no need for an application's custom event types to extend GwtEvent. Prefer `Event` instead.”

We could _possibly_ provide a one-shot `legacy` JAR providing `GwtEvent` and `EventHandler` (and `HasHandlers`?), but people will have to move on to `Event` anyway. IFF we go this route, then I'd rather build this legacy artifact in a separate project: release the JAR once and never maintain it later; lifecycle needs to be separate from the one of gwt-events.

**Comment by jnehlmeier** [comment]
> We could possibly provide a one-shot legacy JAR providing GwtEvent and EventHandler (and HasHandlers?), but people will have to move on to Event anyway. IFF we go this route, then I'd rather build this legacy artifact in a separate project: release the JAR once and never maintain it later; lifecycle needs to be separate from the one of gwt-events.

You could just release gwt-events-1.0 with both classes and being marked deprecated and at the same time gwt-events-2.0 without them. People could jump quicker on the J2CL route and could continue to incrementally modernize their apps. At least that was the general instruction / thinking provided by Colin in gwt-contrib. Of course could also be done with a separate project, although I don't really see a benefit doing so.

I get your point though about all the JavaDoc hints. Not touching such textual notices and actually turn them into real warnings by marking the code deprecated for 7 years now is our fault and we have to live with it now. If we would have marked it explicitly as deprecated so that it would have triggered compile warnings the last 7 years, then it would be fine to just delete it now, but we haven't.

**Comment by FrankHossfeld** [comment]
In c.g.g.evetn.shared we find the EventHandler interface. This marker interface is used for all event handlers (GWT and DOM). You did not migrate this interface. Please, can you clearify why. 

How to deal with it?
* will you add this interface to org.gwtproject.event.shared?
* is it replaced (with what)?
* any other idea how to deal with it?

I would like to hear your suggestions, so that I can solve my problems relating to the dependencies to c,g,g,shared.event.EventHandler

**Comment by niloc132** [comment]
The line I've been trying to take at least is to only remove things actually marked as Deprecated (user.client.Element, DeferredCommand, old Listener wiring, etc). This has never been done for anything in `com.google.gwt.event.`, so for any "finished" code that no one revisits, you'd never see that it is out of date from new deprecation warnings popping up. Plus, there is no automated tooling (unless you're planning on writing some ;) that hunts down "discouraged" packages, though finding all `@Deprecated` in a project and its dependencies is straightforward.

**Comment by tbroyer** [comment]
The way I had seen this is that you first plan for moving to `org.gwtproject`, and possibly move things one part at a time (you can very well move your application event bus without changing your views/widgets at all. And with `gwt-event-compat` you could even move your code in several steps, or move your own code without waiting for third-party libraries to have `org.gwtproject`-compatible releases.)

This **is** going to be a breaking change anyway, and you have to go look for the new names of the packages and modules, and how you replace your `GWT.create()` calls and add the appropriate annotations to trigger the annotation processors. It can be expected that some things disappear (do you really need `Localizable`, `Constants` , `Messages`, `CustomDateTimeFormat`, `SafeHtmlTemplates`, `ClientBundle`, `RemoteService` or other marker interfaces –that are entirely useless in terms of polymorphism– if you have to add annotations anyway to trigger annotation processors?) and you have a little work to do (*replace* the `extends <SomeMarkerInterface>` with an annotation –why keep the marker interface? you'll have to make some change to your interface anyway adding an annotation–, replace the `GWT.create()` call with a `new` of a generate class, or a call to a generated static factory method, and or course change all the `import`s and `<inherits>`)

In the case of events, there's no generator involved, but is that a reason for not having breaking changes other than `import`s and `<inherits>`? Particularly if, in many cases, you can plan for those changes in advance by moving from `c.g.g.event` to `c.g.w.b.event` first.

That being said, I heard @jnehlmeier argument and will add a `gwt-event-legacy` (or `gwt-legacy-event`? we still have a bit of bikeshedding to do on naming) with `GwtEvent` and `EventHandler` marked as `@Deprecated`, and that will only be released once, then deleted right away.

**Comment by FrankHossfeld** [comment]
Can not see, that ```c.g.g.event.shared.EventHandler``` is deprecated. Do I miss something?

**Comment by tbroyer** [comment]
`EventHandler` is useless and only required by the `GwtEvent` API declaration. They go in pair.

**Comment by tbroyer** [comment]
Added in 81e2b1c07bfaa250feded56c0c3daaeb9da5c0ae

**Comment by FrankHossfeld** [comment]
ok, got it. Thanks!
It looks like it is possible to remove the EventHandler interface in gwt-event-dom. Will give it a try.

**Comment by niloc132** [comment]
(Late to the party - didn't finish writing this, but probably the best place the weigh in where all involved parties will see it)

I'd be in favor of changing package but immediately deprecating for "things that no longer make sense, or should have been deprecated long ago". This is going to come up in several areas, like JSO and JSArray in core, plus broken features in gwt-dom, and whatever else we come across...

My general thesis is that rewriting packages across 10s or 1000s of files _and_ requiring other code changes at the same time will prevent projects from even attempting it: If it is more than you can do in intellij's find/replace, it probably won't happen.

GWT.create calls definitely will need their own later pass - here's my thinking:
 * In the "rewrite all packages in my project" pass, leave GWT itself in the old package, since that class is too "magic" to replace outright.
 * Each method will need a replacement of some kind - ones that provide server paths may be problematic without a generalized linker that we can assume all such projects will make use of, and runAsync is an unsolved problem. GWT.create seems actually not terrible - updating can keep the generator (and the other "not-j2cl compatible" wiring) for a version, but all generators should emit a compile-time warning instructing how that type can be updated. Additionally, if the type already conforms to the APT requirements and is already processed, the GWT.create call can note that the class already exists and that the GWT.create call can be replaced by the appropriate constructor or factory call.

**Comment by FrankHossfeld** [comment]
As far as I understood the 'dividing into modules' process, we remove deprecated code, replace JSNI with JsInterop and replace ```com.google.gwt```with ```org.project```.

Is it possible to use ```org.gwtproject.events``` instead of ```org.gwtproject.event.legacy.shared``` as package name for the legacy jar? This will make it easier to transform existing application into the new structure.

I can do a PR. (and add the module descriptor for the legacy module)

**Comment by tbroyer** [comment]
> Is it possible to use org.gwtproject.events instead of org.gwtproject.event.legacy.shared as package name for the legacy jar?

"Split packages" are discouraged in Java 9+ (with JPMS), so I try to play by the (new) rules. That way, if we start adding `Automatic-Module-Name`, we won't break anyone (IIUC, recent versions of maven-compiler-plugin automatically put dependencies on the module path if they are _modules_, i.e. have a `module-info.class` or an `Automatic-Module-Name`).
Even if that artifact is short-lived, I'd rather not do that kind of things; too many cons/risks for almost no pros.

> This will make it easier to transform existing application into the new structure.

It just adds one preparatory step:
```bash
sed -i s/com.google.gwt.event.shared.GwtEvent/org.gwtproject.event.legacy.shared.GwtEvent/g \
    s/com.google.gwt.event.shared.EventHandler/org.gwtproject.event.legacy.shared.EventHandler/g \
    **/*.java
```
before you glob the whole package:
```bash
sed -i s/com.google.gwt.event/org.gwtproject.event/g \
    s/com.google.web.bindery.event/org.gwtproject.event/g \
    **/*.java
```
(or fixing it afterwards if you skipped the preparatory step and went straight to the package glob replace:
```bash
sed -i s/org.gwtproject.event.shared.GwtEvent/org.gwtproject.event.legacy.shared.GwtEvent/g \
    s/org.gwtproject.event.shared.EventHandler/org.gwtproject.event.legacy.shared.EventHandler/g \
    **/*.java
```
)

**Comment by jnehlmeier** [comment]
@tbroyer do you mind adding HandlerManager to the legacy jar as well? I am working on gwt-widgets and noticed that it is needed as well. Widget has a protected method to create a HandlerManager and allows sub classes to override that method.

**Comment by tbroyer** [comment]
How about adding `HandlerManager` in gwt-widgets then? (yes, that would make one more special case when batch-replacing imports, but `HandlerManager` is a) widget-specific, b) not deprecated)

Also, note that I had removed support for firing events in reverse order here. AFAICT this was only used by [`Event.addNativePreviewHandler`](https://github.com/gwtproject/gwt/blob/a90832fec9a56f4cb83ed009c7048e0a8d196fe8/user/src/com/google/gwt/user/client/Event.java#L458), so it'll have to be reimplemented.

**Comment by FrankHossfeld** [comment]
@tbroyer: Is the module descriptor missing in the legacy module?

**Comment by jnehlmeier** [comment]
@FrankHossfeld I guess the descriptor is accidentally in `/src/test/java` instead of `/src/main/java`.

@tbroyer Hmm looks like most occurrences of `HandlerManager` can be replaced with `SimpleEventBus` today without breaking anybody (private use), except in `Widget` itself (sub classes) and because of reverse ordering in `Event.addNativePreviewHandler()`. Do you know any rational why handlers are notified in reverse ordering? 

Personally I would mark `HandlerManager` `@deprecated` as per JavaDoc the use of `HandlerManager` is highly discouraged. In a later release then replace it with `EventBus`. Thats why I thought it might be a good idea to add it to your legacy library.

**Comment by tbroyer** [comment]
>  I guess the descriptor is accidentally in /src/test/java instead of /src/main/java.

Oops, will fix.

> Hmm looks like most occurrences of HandlerManager can be replaced with SimpleEventBus

**Only** if you create it with a `null` source.

Let's "force" everyone who still uses it as an event bus to make the switch to `SimpleEventBus` (or import gwt-widgets).

> Do you know any rational why handlers are notified in reverse ordering?

IIRC, this is explained in the javadoc :wink:

**Comment by tbroyer** [comment]
BTW `addNativePreviewHandler` could easily be replaced with a simple stack, as there's only one event type (no need for the dispatch-by-type or even filter-by-source of the event bus / handler manager)

**Comment by jnehlmeier** [comment]
> IIRC, this is explained in the javadoc 😉

Hmm I have actually read some JavaDoc here and there but it doesn't really explain why. [It only says that they want to fire it in reverse order](https://github.com/gwtproject/gwt/blob/master/user/src/com/google/gwt/user/client/Event.java#L403) and [explains to users that it will fire in reverse order](https://github.com/gwtproject/gwt/blob/master/user/src/com/google/gwt/user/client/Event.java#L435). Maybe I haven't yet found the JavaDoc explaining it.

**Comment by tbroyer** [comment]
Ah, right. Event exposes mutable isConsumed and isCancelled, so order matters, and we want a stack for things like popups. Event also exposes isFirstHandler, which is used in popup panel.

(Love the doc on Event.handlers :wink:)

**Pull request closed without merging**